src/string.cr

```diff
@@ -522,18 +522,19 @@ class String
     end
 
     negative = false
-    found_digit = false
-    mul_overflow = ~0_u64 / base
 
     # Check + and -
     case ptr.value.unsafe_chr
-    when '+'
-      ptr += 1
     when '-'
+      return ToU64Info.new 0, true, true if unsigned
```

Comment:
The other little optimizations are probably worth the change, but I'm questioning this one. Parsing negative numbers as unsigned integer should be a very, very rare case. UInts are usually only used in binary protocols anyway.

This line gives a nice performance boost for this rare case 10 times. But it also affects the other cases. It just costs a little performance, but parsing a negative number as a signed integer is a much more common scenario.

Reply:
> Parsing negative numbers as unsigned integer should be a very, very rare case. UInts are usually only used in binary protocols anyway.

I actually have a case in my code. I'm converting a string to an integer and the integer must not be negative so I simply do: `string.to_u32?` which will obviously also return `nil` when the integer in the string is negative. I don't think cases where you don't want to allow negative numbers are that rare.

Most of the time in the benchmarks the new `to_i` methods were `1.04× slower` or less slower for negative numbers but it's actually quite easy to get the runtime `if unsigned` check over to compile-time by simply making `to_u64_info` a macro and putting the `if unsigned` check in a macro branch.
So now there are no longer trade-offs.

Reply:
> I don't think cases where you don't want to allow negative numbers are that rare.

Yeah, but such cases would usually parse a signed integer and check for `>= 0`. Using an unsigned integer type for this is a creative solution but should not really be recommended.

> So now there are no longer trade-offs.

Maybe, but I don't see a relevant benefit either.

Reply:
> Maybe, but I don't see a relevant benefit either.

But you said this previously:

> And it also affects the other cases. It just costs a little performance, but parsing a negative number as a signed integer is a much more common scenario. I wouldn't want to sacrifice this for a doubtful improvement.

Didn't you say by this that the `if unsigned` runtime check could cost perfomance for parsing negative signed integers? That's what I resolved with the macro. The `if unsigned` check no longer costs any perfomance for parsing negative numbers as signed integers. It no longer affects any other cases.

Reply:
That was *a* point. Others are general complexity, increased compile time, resources needed for implementing, reviewing and applying this patch, successive maintenance etc. All for an IMHO very marginal benefit.

Reply:
I don't oppose this PR if others think it's worth merging. But I won't approve it either.